include/nbl/ext/TextRendering/TextRendering.h

```diff
@@ -0,0 +1,97 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_EXT_TEXT_RENDERING_H_INCLUDED_
+#define _NBL_EXT_TEXT_RENDERING_H_INCLUDED_
+
+#include "nabla.h"
+
+#include "nbl/video/alloc/SubAllocatedDataBuffer.h"
+#include "nbl/video/utilities/CPropertyPool.h"
+
+using namespace nbl;
+using namespace nbl::core;
+using namespace nbl::asset;
+using namespace nbl::video;
+
+namespace nbl
+{
+namespace ext
+{
+namespace TextRendering
+{
+
+struct SPixelCoord
+{
+	uint16_t x, y;
+};
+
+struct SGlyphData 
+{
+	SPixelCoord offsetFromStringMin;
+	uint32_t textureAtlasGlyphIndex;
+};
+
+struct StringBoundingBox
+{
+	SPixelCoord min, max;
+};
+
+class NBL_API TextRenderer
+{
+public:
+	TextRenderer(core::smart_refctd_ptr<ILogicalDevice> device, uint32_t maxGlyphCount, uint32_t maxStringCount);
+
+	bool allocateString(
+		// Offset in pixels from top left in the screen
+		int offsetX, int offsetY,
+		uint32_t glyphCount,
+		SGlyphData const* glyphs,
+		// Offset into allocated parts of m_stringDataPropertyPool & m_geomDataBuffer
+		uint32_t* outStringOffset,
+		uint32_t* outGeometryOffset
+	);
```

Comment:
have this signature instead
```cpp
using glyph_geometry_pool_t = video::SubAllocatedDataBufferST<>;
using string_pool_t = video::CPropertyPool<core::allocator,glyph_geometry_pool_t::size_type,StringBoundingBox>;

struct string_handle_t
{
    string_pool_t::size_type stringAddr = string_pool_t::invalid_address;
    glyph_geometry_pool_t::size_type glyphDataAddr = glyph_geometry_pool_t::invalid_address;
    glyph_geometry_pool_t::size_type glyphCount = 0u;
};

// how many were successfully allocated, `handles` will be properly filled between [0,retval)
uint32_t allocateStrings(
    const std::chrono::time_point<Clock>& maxWaitPoint, // lets have also a version without this (default)
    const uint32_t count, // how many strings
    string_handle_t* handles, // output handles, if `glyphDataAddr` was not primed with invalid_address, allocation will not happen, likewise for `stringDataAddr`
    const char* const* stringData,
    const StringBoundingBox* wrappingBoxes = nullptr // optional, to wrap paragraphs
);

void freeStrings(
    const uint32_t count, // how many strings
    const string_handle_t* handles,
    core::smart_refctd_ptr<IGPUFence>&& fence // for a deferred free
);
```

Then if you could use your custom DeferredFree functor
https://github.com/Devsh-Graphics-Programming/Nabla/blob/518ef10fc445326c9b9274216bc5800545ab2a8d/include/nbl/video/alloc/SubAllocatedDataBuffer.h#L61
that would also free the string addresses